### DISCUSSION TITLE
WorkQueue if watch two update events of the same resource in the short time, it how to enqueue

### DISCUSSION THREAD

**Discussion by VictoryAnn** [kubernetes-sigs/kubebuilder#2469] [Support] [answered]
1. Question: I update the same CR twice in the short time, and different workers get key from work queue, will it happen that two workers reconcile different events on the same CR ?  How to ensure that the new event reconcile finished finally  instead of the old one reconcile finished finally?
1. Puzzled： if watch two update events of the same resource in the short time, how it enqueue.  
2. Solution:   Need I use lock to prevent two workers reconcile the same CR ?

**Comment by camilamacedo86** [comment, marked as answer]
First at all, see that the golden rule is that your solutions should be idempotent. Therefore, the order should not matter. Then, by doing idempotent solutions you can trust the reconciliation process. It's designed to handle such scenarios gracefully. 

For further details

- **Order of Processing:** The controller-runtime's reconciliation mechanism does not guarantee the order of processing when multiple updates to a resource occur rapidly. Especially with multiple workers in play, it's possible for them to pick up the newer event before the older one.
- **Reconciliation:** Kubebuilder/controllers designed using controller-runtime follow a level-based approach rather than an edge-based one. The primary aim of the reconciliation logic is to ensure the observed state aligns with the desired state, irrespective of the order of events. Thus, even if a newer event is processed before an older one, the final state of the resource should reflect the desired state as per the latest update.
- **Managing Concurrency:** controller-runtime allows for concurrent reconciliations, but it ensures that the same resource isn't processed concurrently by multiple workers. This is managed internally, so you typically don't need to implement locking mechanisms manually. If you observe issues, it may be worthwhile to inspect if there are external factors influencing the order of operations or if there's a specific logic within the reconciliation that might be causing unexpected behaviors.

I hope that answered your question
I am closing this one as answered.
However, if you think that should be re-opened please feel free to do so.
